File: compiler/bench/unrollable2.c
// expected: 823739198976
// Polynomial evaluation unroll pattern - Horner's method
#include <stdio.h>

long poly(long x) {
    return 1 + x + 2*x*x + 3*x*x*x;
}

int main() {
    long sum = 0;
    for (int i = 0; i < 1024; i++) sum += poly(i);
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/unrollable2.fs
\ expected: 823739198976
\ Polynomial evaluation unroll pattern - Horner's method

: poly ( x -- result )
  dup dup * >r     \ x; r: x^2
  dup r@ * >r      \ x; r: x^2 x^3
  r> 3 * swap      \ 3*x^3 x
  r> 2 * +         \ 3*x^3+2*x^2 x
  + 1+ ;           \ 3*x^3+2*x^2+x+1

: main
  0 1024 0 do i poly + loop . cr ;
